src/libs/kata-types/src/config/hypervisor/qemu.rs

```diff
@@ -0,0 +1,146 @@
+    pub fn new() -> Self {
+        QemuConfig {}
+    }
+
+    /// Register the qemu plugin.
+    pub fn register(self) {
+        let plugin = Arc::new(self);
+        register_hypervisor_plugin(HYPERVISOR_NAME_QEMU, plugin);
+    }
+}
+
+impl ConfigPlugin for QemuConfig {
+    fn get_max_cpus(&self) -> u32 {
+        MAX_QEMU_VCPUS
+    }
+
+    fn get_min_memory(&self) -> u32 {
+        MIN_QEMU_MEMORY_SIZE
+    }
+    fn name(&self) -> &str {
+        HYPERVISOR_NAME_QEMU
+    }
+
+    /// Adjust the configuration information after loading from configuration file.
+    fn adjust_configuration(&self, conf: &mut TomlConfig) -> Result<()> {
+        if let Some(qemu) = conf.hypervisor.get_mut(HYPERVISOR_NAME_QEMU) {
+            if qemu.path.is_empty() {
+                qemu.path = default::DEFAULT_QEMU_BINARY_PATH.to_string();
+            }
+            resolve_path!(qemu.path, "Qemu binary path `{}` is invalid: {}")?;
+            if qemu.ctlpath.is_empty() {
+                qemu.ctlpath = default::DEFAULT_QEMU_CONTROL_PATH.to_string();
+            }
+            resolve_path!(qemu.ctlpath, "Qemu ctlpath `{}` is invalid: {}")?;
+
+            if qemu.boot_info.kernel.is_empty() {
+                qemu.boot_info.kernel = default::DEFAULT_QEMU_GUEST_KENREL_IMAGE.to_string();
+            }
+            if qemu.boot_info.kernel_params.is_empty() {
+                qemu.boot_info.kernel_params =
+                    default::DEFAULT_QEMU_GUEST_KENREL_PARAMS.to_string();
+            }
+            if qemu.boot_info.firmware.is_empty() {
+                qemu.boot_info.firmware = default::DEFAULT_QEMU_FIRMWARE_PATH.to_string();
+            }
+
+            if qemu.device_info.default_bridges == 0 {
+                qemu.device_info.default_bridges = default::DEFAULT_QEMU_PCI_BRIDGES;
+                if qemu.device_info.default_bridges > default::MAX_QEMU_PCI_BRIDGES {
+                    qemu.device_info.default_bridges = default::MAX_QEMU_PCI_BRIDGES;
+                }
+            }
+
+            if qemu.machine_info.machine_type.is_empty() {
+                qemu.machine_info.machine_type = default::DEFAULT_QEMU_MACHINE_TYPE.to_string();
+            }
+            if qemu.machine_info.entropy_source.is_empty() {
+                qemu.machine_info.entropy_source = default::DEFAULT_QEMU_ENTROPY_SOURCE.to_string();
+            }
+
+            if qemu.memory_info.default_memory == 0 {
+                qemu.memory_info.default_memory = default::DEFAULT_QEMU_MEMORY_SIZE;
+            }
+            if qemu.memory_info.memory_slots == 0 {
+                qemu.memory_info.memory_slots = default::DEFAULT_QEMU_MEMORY_SLOTS;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Validate the configuration information.
+    fn validate(&self, conf: &TomlConfig) -> Result<()> {
+        if let Some(qemu) = conf.hypervisor.get(HYPERVISOR_NAME_QEMU) {
+            validate_path!(qemu.path, "Qemu binary path `{}` is invalid: {}")?;
```

Comment:
Nits for the error messages: The official spelling is `QEMU`.

Reply:
Fixed